File: 1617-stone-game-iv/stone-game-iv.java
class Solution {
    public boolean winnerSquareGame(int n) {
        Boolean[][] dp=new Boolean[n+1][2];
        return winner(n,true,dp);
    }
    public boolean winner(int n,boolean turn,Boolean[][] dp){
        if(n==0){
            return false;
        }
        int t=turn ? 1 :0;
        if(dp[n][t]!=null){
            return dp[n][t];
        }
        double x=Math.sqrt(n);
        boolean f1=false;
        boolean f2=false;
        if(turn){
            if((int)x*(int)x==n){
                dp[n][t]=true;
                return true;
            }
            else{
                for(int k=1;k<=(int)x;k++){
                    if(!winner(n-k*k,false,dp)){
                        f1=true;
                        break;
                    }
                }
            }
            dp[n][t]=f1;
            return f1;
        }
        else{
            if((int)x*(int)x==n){
                dp[n][t]=true;
                return true;
            }
            else{
                for(int k=1;k<=(int)x;k++){
                    if(!winner(n-k*k,true,dp)){
                        f2=true;
                        break;
                    }
                }
            }
            dp[n][t]=f2;
            return f2;
        }     
    }
}

File: 1617-stone-game-iv/stone-game-iv.cpp
class Solution {
public:
    vector<vector<int>> dp; 
    bool winnerSquareGame(int n) {
        dp.assign(n + 1, vector<int>(2, -1));
        return solve(n, 1);
    }

    bool solve(int n, int turn) {
        if(n == 0){
            return false;
        }
        int x=sqrt(n);
        if(dp[n][turn]!=-1){
            return dp[n][turn];
        }
        if(x * x == n){
            return true;
        }

        if (turn) {
            for (int i = 1; i * i <= n; i++) {
                if (!solve(n - i * i, 0)){//previus move of alice is winning
                    return dp[n][turn]=true;
                }
            }
            return dp[n][turn]=false;
        } 
        else {
            for (int i = 1; i * i <= n; i++) {
                if (!solve(n - i * i, 1))
                    return dp[n][turn]=true;
            }
            return dp[n][turn]=false;
        }
    }
};